src/object/srv_ec_aggregate.c

```diff
@@ -649,17 +1642,56 @@ agg_process_stripe(struct ec_agg_entry *entry)
 		update_vos = false;
 		goto out;
 	}
+	/* Parity, some later replicas, possibly holes, not full stripe. */
+	if (entry->ae_cur_stripe.as_has_holes)
```

Comment:
you only need to process the holes that with higher epoch than parity's epoch right?
if hole's epoch < parity epoch, can just remove those records. 

Above L1621 "entry->ae_par_extent.ape_epoch >= entry->ae_cur_stripe.as_hi_epoch" that comparison seems not enough, because it is possibly with both replica ext (with higher epoch) and hole ext (with lower epoch).
If all holes are all with lower epoch than parity epoch, need not agg_process_holes(), no?

Reply:
You're right here.  I will do a subsequent patch that does these fixes and adds tests for range punch.